src/charm.py

```diff
@@ -328,30 +394,46 @@ def _reconcile_runners(self, runner_manager: RunnerManager) -> Dict[str, "JsonOb
         Returns:
             Changes in runner number due to reconciling runners.
         """
-        virtual_machines_resources = VMResources(
+        virtual_machines_resources = VirtualMachineResources(
             self.config["vm-cpu"], self.config["vm-memory"], self.config["vm-disk"]
         )
-        # handle deprecated config for `quantity` and `virt-type`
-
-        containers = self.config["containers"]
-        if containers == 0 and self.config["virt-type"] == "container":
-            containers = self.config["quantity"]
 
         virtual_machines = self.config["virtual-machines"]
         if virtual_machines == 0 and self.config["virt-type"] == "virtual-machine":
             virtual_machines = self.config["quantity"]
 
-        delta_containers = runner_manager.reconcile("container", containers)
         delta_virtual_machines = runner_manager.reconcile(
-            "virtual-machine", virtual_machines, virtual_machines_resources
+            virtual_machines, virtual_machines_resources
         )
-        return {
-            "delta": {
-                "containers": delta_containers,
-                "virtual-machines": delta_virtual_machines,
-            }
-        }
+        return {"delta": {"virtual-machines": delta_virtual_machines}}
+
+    @staticmethod
+    @retry(tries=10, delay=15, max_delay=60, backoff=1.5)
+    def _install_deps() -> None:
+        """Install dependencies."""
+        logger.info("Installing charm dependencies.")
+
+        # Binding for snap, apt, and lxd init commands are not available so subprocess.run used.
+        execute_command(["/usr/bin/apt", "remove", "-qy", "lxd", "lxd-client"], check=False)
+        execute_command(["/usr/bin/snap", "install", "lxd", "--channel=latest/stable"])
```

Comment:
For any of these commands, is there any issues with running them multiple times? E.g., a later command fails which triggers the retry which then leads to these commands being run again and they fail? Or are they idempotent?

Reply:
I reviewed them, and they all seem idempotent.